File: server/internal/wealth/db/asset_adapter_sources.go
package wealthdb

import (
	"context"
	"database/sql"
	"errors"
	"fmt"

	"github.com/samber/lo"

	"tallyo/internal/database/dbgen"
	"tallyo/internal/graph/model"
	"tallyo/internal/wealth"
	"tallyo/internal/wealth/syncerids"
)

func (s *Store) AssetAdapterSourcesByAssetIDs(ctx context.Context, assetIDs []int64) (map[int64][]*model.AssetAdapterSource, error) {
	rows, err := s.q.AssetAdapterSourcesByAssetIDs(ctx, dbgen.AssetAdapterSourcesByAssetIDsParams{AssetIds: assetIDs})
	if err != nil {
		return nil, fmt.Errorf("list asset adapter sources: %w", err)
	}

	toSourceByAssetID := func(row dbgen.AssetAdapterSourcesByAssetIDsRow) (int64, *model.AssetAdapterSource) {
		return row.AssetID, &model.AssetAdapterSource{
			SourceAdapter: wealth.SourceAdapterFromSyncerID(syncerids.ID(row.SourceAdapter)),
			SourceID:      row.SourceID,
		}
	}
	return lo.GroupByMap(rows, toSourceByAssetID), nil
}

func (s *Store) AssetByAdapterSource(ctx context.Context, source wealth.AdapterSource) (*model.Asset, error) {
	assetID, err := s.q.AssetAdapterSourceAssetID(ctx, dbgen.AssetAdapterSourceAssetIDParams{
		SourceAdapter: source.Adapter.Str(),
		SourceID:      source.SourceID,
	})
	if errors.Is(err, sql.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, fmt.Errorf("lookup adapter source: %w", err)
	}
	return assetByID(ctx, s.q, assetID)
}

func (s *Store) MergeAssetBySource(ctx context.Context, adapter syncerids.ID, sourceID string, targetAssetID int64) (*model.Asset, error) {
	var result *model.Asset
	if err := s.WithTx(ctx, func(_ *sql.Tx, q *dbgen.Queries) error {
		duplicateAssetID, err := q.AssetAdapterSourceAssetID(ctx, dbgen.AssetAdapterSourceAssetIDParams{
			SourceAdapter: adapter.Str(),
			SourceID:      sourceID,
		})
		if errors.Is(err, sql.ErrNoRows) {
			return fmt.Errorf("no such adapter source %s:%s", adapter, sourceID)
		}
		if err != nil {
			return fmt.Errorf("lookup adapter source: %w", err)
		}
		if duplicateAssetID == targetAssetID {
			result, err = assetByID(ctx, q, targetAssetID)
			return err
		}

		duplicate, err := assetByID(ctx, q, duplicateAssetID)
		if err != nil {
			return err
		}
		if duplicate == nil {
			return fmt.Errorf("asset %d not found", duplicateAssetID)
		}

		target, err := assetByID(ctx, q, targetAssetID)
		if err != nil {
			return err
		}
		if target == nil {
			return fmt.Errorf("asset %d not found", targetAssetID)
		}
		if duplicate.AssetType != target.AssetType {
			return fmt.Errorf("cannot merge %s asset %d into %s asset %d", duplicate.AssetType, duplicateAssetID, target.AssetType, targetAssetID)
		}

		hasCollision, err := q.AssetMergeCollisionExists(ctx, dbgen.AssetMergeCollisionExistsParams{
			DuplicateAssetID: duplicateAssetID,
			TargetAssetID:    targetAssetID,
		})
		if err != nil {
			return fmt.Errorf("check asset merge collision: %w", err)
		}
		if hasCollision {
			return fmt.Errorf("cannot merge assets %d and %d because a snapshot or account/date holds both", duplicateAssetID, targetAssetID)
		}

		params := dbgen.UpdateAssetDailyHoldingsAssetIDParams{
			DuplicateAssetID: duplicateAssetID,
			TargetAssetID:    targetAssetID,
		}
		if err := q.UpdateAssetDailyHoldingsAssetID(ctx, params); err != nil {
			return fmt.Errorf("move asset holdings: %w", err)
		}
		if err := q.MoveAssetAnalysisReport(ctx, dbgen.MoveAssetAnalysisReportParams(params)); err != nil {
			return fmt.Errorf("move asset analysis report: %w", err)
		}
		if err := q.MoveAssetAdapterSources(ctx, dbgen.MoveAssetAdapterSourcesParams(params)); err != nil {
			return fmt.Errorf("move asset adapter sources: %w", err)
		}
		if err := q.DeleteAssetByID(ctx, dbgen.DeleteAssetByIDParams{ID: duplicateAssetID}); err != nil {
			return fmt.Errorf("delete duplicate asset: %w", err)
		}

		result = target
		return nil
	}); err != nil {
		return nil, err
	}
	return result, nil
}

File: server-rs/src/wealth/store/asset_adapter_sources.rs
use anyhow::{Result, ensure};
use sqlx::{Executor, Sqlite, SqlitePool};

use crate::{
    database::queries,
    wealth::{AdapterSource, Asset},
};

use super::assets::asset_by_id;

pub async fn asset_by_adapter_source(pool: &SqlitePool, source: &AdapterSource) -> Result<Option<Asset>> {
    let adapter = source.adapter.to_string();
    let Some(row) = queries::asset_adapter_source_asset_id_opt(
        pool,
        queries::AssetAdapterSourceAssetIdParams {
            source_adapter: &adapter,
            source_id: &source.source_id,
        },
    )
    .await?
    else {
        return Ok(None);
    };
    asset_by_id(pool, row.asset_id).await
}

pub async fn validate_asset_merge(
    executor: impl Executor<'_, Database = Sqlite>,
    duplicate_asset_id: i64,
    target_asset_id: i64,
) -> Result<()> {
    ensure!(duplicate_asset_id != target_asset_id, "asset IDs must differ");
    let collision = queries::asset_merge_collision_exists(
        executor,
        queries::AssetMergeCollisionExistsParams {
            duplicate_asset_id,
            target_asset_id,
        },
    )
    .await?
    .has_collision;
    ensure!(
        !collision,
        "cannot merge assets {duplicate_asset_id} and {target_asset_id} because a snapshot or account/date holds both"
    );
    Ok(())
}

#[cfg(test)]
mod tests {
    use anyhow::Result;

    use super::{asset_by_adapter_source, validate_asset_merge};
    use crate::{
        database::dbtest,
        wealth::{AdapterSource, SyncerId},
    };

    #[tokio::test]
    async fn finds_adapter_assets_and_rejects_self_merges() -> Result<()> {
        let pool = dbtest::open().await?;
        let asset_id = sqlx::query_scalar::<_, i64>(
            "INSERT INTO assets (asset_type, identifier, classifier) VALUES ('CRYPTO', 'ETH', 'CRYPTOCURRENCY') RETURNING id",
        )
        .fetch_one(&pool)
        .await?;
        sqlx::query(
            "INSERT INTO asset_adapter_sources (asset_id, source_adapter, source_id) VALUES (?, 'debank', '0xeth')",
        )
        .bind(asset_id)
        .execute(&pool)
        .await?;
        let source = AdapterSource {
            adapter: SyncerId::Debank,
            source_id: "0xeth".to_owned(),
        };

        assert_eq!(
            asset_by_adapter_source(&pool, &source).await?.map(|asset| asset.id),
            Some(asset_id)
        );
        assert!(
            asset_by_adapter_source(
                &pool,
                &AdapterSource {
                    source_id: "missing".to_owned(),
                    ..source
                }
            )
            .await?
            .is_none()
        );
        assert_eq!(
            validate_asset_merge(&pool, asset_id, asset_id)
                .await
                .unwrap_err()
                .to_string(),
            "asset IDs must differ"
        );
        validate_asset_merge(&pool, asset_id, asset_id + 1).await?;
        Ok(())
    }
}
